without-webhooks/client/ios-swift/app/CheckoutViewController.swift

```diff
@@ -0,0 +1,219 @@
+                }
+            }
+            // Payment succeeded
+            else if clientSecret != nil && (requiresAction == nil || requiresAction == false) {
+                STPAPIClient.shared().retrievePaymentIntent(withClientSecret: clientSecret!) { paymentIntent, retrieveError in
+                    DispatchQueue.main.async {
+                        if let paymentIntent = paymentIntent {
+                            let alert = UIAlertController(title: "Payment succeeded", message: paymentIntent.description, preferredStyle: .alert)
+                            alert.addAction(UIAlertAction(title: "Restart demo", style: .cancel) { _ in
+                                self?.cardTextField.clear()
+                                self?.loadPage()
+                            })
+                            self?.present(alert, animated: true, completion: nil)
+                        }
+                        else {
+                            let message = retrieveError?.localizedDescription ?? ""
+                            let alert = UIAlertController(title: "Payment failed", message: message, preferredStyle: .alert)
+                            alert.addAction(UIAlertAction(title: "OK", style: .cancel))
+                            self?.present(alert, animated: true, completion: nil)
+                        }
+                    }
+                }
+            }
+            // Payment requires additional actions
+            else if clientSecret != nil && requiresAction == true && self != nil {
+                let paymentHandler = STPPaymentHandler.shared()
+                paymentHandler.handleNextAction(forPayment: clientSecret!, authenticationContext: self!, returnURL: nil) { status, paymentIntent, handleActionError in
+                    DispatchQueue.main.async {
+                        switch (status) {
+                        case .failed:
+                            let message = handleActionError?.localizedDescription ?? ""
+                            let alert = UIAlertController(title: "Payment failed", message: message, preferredStyle: .alert)
+                            alert.addAction(UIAlertAction(title: "OK", style: .cancel))
+                            self?.present(alert, animated: true, completion: nil)
+                            break
+                        case .canceled:
+                            let message = handleActionError?.localizedDescription ?? ""
+                            let alert = UIAlertController(title: "Payment canceled", message: message, preferredStyle: .alert)
+                            alert.addAction(UIAlertAction(title: "OK", style: .cancel))
+                            self?.present(alert, animated: true, completion: nil)
+                            break
+                        case .succeeded:
+                            if let paymentIntent = paymentIntent, paymentIntent.status == STPPaymentIntentStatus.requiresConfirmation {
```

Comment:
We should update the copy here, and consider renaming methods to make it clearer why this step is necessary.

We can reframe “you have to re-confirm again” as:
“This second confirm step gives you the opportunity to fulfill the payment synchronously on the server”

We should probably also reframe this in the server code.

Reply:
(added comment)